pkg/cmd/admin/migrate/etcd/ttl.go

```diff
@@ -0,0 +1,162 @@
+		Short:   "Attach keys to etcd v3 leases to assist in etcd v2 migrations",
+		Long:    internalMigrateTTLLong,
+		Example: fmt.Sprintf(internalMigrateTTLExample, fullName),
+		Run: func(cmd *cobra.Command, args []string) {
+			kcmdutil.CheckErr(options.Run())
+		},
+	}
+
+	options.Bind(cmd.Flags())
+
+	return cmd
+}
+
+func (o *MigrateTTLReferenceOptions) Bind(flag *pflag.FlagSet) {
+	flag.StringVar(&o.etcdAddress, "etcd-address", "", "Etcd address")
+	flag.StringVar(&o.ttlKeysPrefix, "ttl-keys-prefix", "", "Prefix for TTL keys")
+	flag.DurationVar(&o.leaseDuration, "lease-duration", time.Hour, "Lease duration (seconds granularity)")
+	flag.StringVar(&o.certFile, "cert", "", "identify secure client using this TLS certificate file")
+	flag.StringVar(&o.keyFile, "key", "", "identify secure client using this TLS key file")
+	flag.StringVar(&o.caFile, "cacert", "", "verify certificates of TLS-enabled secure servers using this CA bundle")
+}
+
+func generateClientConfig(o *MigrateTTLReferenceOptions) (*clientv3.Config, error) {
+	if o.etcdAddress == "" {
+		return nil, fmt.Errorf("--etcd-address flag is required")
+	}
+	if o.ttlKeysPrefix == "" {
+		return nil, fmt.Errorf("--ttl-keys-prefix flag is required")
+	}
+
+	c := &clientv3.Config{
+		Endpoints: []string{o.etcdAddress},
+	}
+
+	var cfgtls *transport.TLSInfo
+	tlsinfo := transport.TLSInfo{}
+	if o.certFile != "" {
+		tlsinfo.CertFile = o.certFile
+		cfgtls = &tlsinfo
+	}
+
+	if o.keyFile != "" {
+		tlsinfo.KeyFile = o.keyFile
+		cfgtls = &tlsinfo
+	}
+
+	if o.caFile != "" {
+		tlsinfo.CAFile = o.caFile
+		cfgtls = &tlsinfo
+	}
+
+	if cfgtls != nil {
+		glog.V(4).Infof("TLS configuration: %#v", cfgtls)
+		clientTLS, err := cfgtls.ClientConfig()
+		if err != nil {
+			return nil, err
+		}
+		c.TLS = clientTLS
+	}
+	return c, nil
+}
+
+func (o *MigrateTTLReferenceOptions) Run() error {
+	c, err := generateClientConfig(o)
+	if err != nil {
+		return err
+	}
+	glog.V(4).Infof("Using client config: %#v", c)
+
+	client, err := clientv3.New(*c)
+	if err != nil {
+		return fmt.Errorf("unable to create etcd client: %v", err)
+	}
+
+	// Make sure that ttlKeysPrefix is ended with "/" so that we only get children "directories".
+	if !strings.HasSuffix(o.ttlKeysPrefix, "/") {
+		o.ttlKeysPrefix += "/"
+	}
+	ctx := context.Background()
+
+	objectsResp, err := client.KV.Get(ctx, o.ttlKeysPrefix, clientv3.WithPrefix())
+	if err != nil {
+		return fmt.Errorf("unable to get objects to attach to the lease: %v", err)
+	}
+
+	lease, err := client.Lease.Grant(ctx, int64(o.leaseDuration/time.Second))
+	if err != nil {
+		return fmt.Errorf("unable to create lease: %v", err)
+	}
+	fmt.Fprintf(o.Out, "info: Lease with TTL %d created\n", lease.TTL)
+
+	fmt.Fprintf(o.Out, "info: Attaching lease to %d entries\n", len(objectsResp.Kvs))
+	errors := 0
+	for _, kv := range objectsResp.Kvs {
+		_, err := client.KV.Put(ctx, string(kv.Key), string(kv.Value), clientv3.WithLease(lease.ID))
```

Comment:
make conditional so we don't stomp a changed value, gather items we need to retry?